### PR TITLE
docs: add rust dev environment with nix and direnv

### PR DESCRIPTION
**Pull request by GTrunSec**
@ericdallo  May you help me to modify and enrich this doc. thanks
@DavidDeSimone please check this doc that pretty simple.

### CODE REVIEW

File: docs/nix-develop.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# Using Nix Develop Environment
```
Comment (ericdallo) [inline docs/nix-develop.md:1]: How users would access this content? we can add a new entry on `mkdocs.yaml` or reference it from an existing docs file or both

File: docs/nix-develop.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+## Install nix from shell
+
+    sh <(curl -L https://nixos.org/nix/install) --daemon
+
+then `ecec bash -l` reload your bash to load the nix-env
```
Comment (ericdallo) [inline docs/nix-develop.md:7]:
```suggestion
then `exec bash -l` reload your bash to load the nix-env
```

File: docs/nix-develop.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
+# Using Nix Develop Environment
+
```
Comment (ericdallo) [inline docs/nix-develop.md:2]: Maybe we can detail here why we are using nix to build/develop, a brief summary with the advantages of nix?

File: docs/nix-develop.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+1. Night version
+   - located `nix/rust.nix` modify `2021-01-14`to your want.([rustOverlay-NightlyCheck](https://github.com/oxalica/rust-overlay/tree/master/manifests/nightly/default.nix)) 
+     Example: `default = pkgs.rust-bin.nightly."2021-03-23";`
+
+2. stable version
+
+ - example : `default = pkgs.rust-bin.stable."1.50.0";`
+   ([rustOverlay-StableCheck](https://github.com/oxalica/rust-overlay/tree/master/manifests/stable/default.nix)) 
+
+3. beat version
```
Comment (ericdallo) [inline docs/nix-develop.md:41]:
```suggestion
3. beta version
```

File: docs/nix-develop.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      check valid package from https://search.nixos.org/packages?channel=unstable";
+      ";
+    };
+```
+
+### Add nixpkgs package to current Environment
+
+- located `nix/rust.nix`
+
+first search any package in [search-package](https://search.nixos.org/packages?channel=unstable&) then add the name to
```
Comment (ericdallo) [inline docs/nix-develop.md:85]:
```suggestion
First search any package in [search-package](https://search.nixos.org/packages?channel=unstable&) then add the name to:
```

File: docs/nix-develop.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      ++ map (tool: cfg.rustOverlay.${tool}) cfg.rustOverlaySet ++ (with pkgs;[
+      #custom nixpkgs packages
+      rustracer
+      <add package here>
+    ]);
+ ```
+
+## Reload all of envs when you changed something
+
+- normally, we can re-enter `nix-shell` to reload envrs. for convinet
```
Comment (ericdallo) [inline docs/nix-develop.md:98]: `for convinet`
I didn't get this part

File: docs/nix-develop.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+```toml
+[[env]]
+name = "TEST"
+vale = "/bin/test"
+#prefix = "$( cd "$(dirname "$\{\BASH_SOURCE [ 0 ]}")"; pwd )" can be prefix 
+```
+
+
+
```
Comment (ericdallo) [inline docs/nix-develop.md:161]:
```suggestion
```

Comment (GTrunSec) [reply docs/nix-develop.md:161]: Sorry, wait a moment. Let me rewrite the doc.

**Review by ericdallo — COMMENTED** [review]
Good job!

The `nix-develop` section looks a bit too verbose to me, but I don't have rust context, also I'm not a native speaker but I think we could improve the wording. Maybe @DavidDeSimone or @brotzeit could take a look as they input as non nix users could help

File: docs/nix-develop.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+## Clone Emacs-ng by nix-shell and enable emacs cachix 
+
+```bash
+nix-env -iA cachix -f https://cachix.org/api/v1/install
+exec bash -l
+cachix use emacsng # make sure you have saw the output like: Configured https://emacsng.cachix.org binary cache in /home/test/.config/nix/nix.conf
+nix-shell -p git --command "git clone https://github.com/emacs-ng/emacs-ng.git && cd emacs-ng && nix-shell"
+```
+
+## Setting up Rust develop environment
```
Comment (brotzeit) [inline docs/nix-develop.md:29]: Is this all relevant in order to install emacs-ng ?

**Comment by brotzeit** [comment]
I think it would be cool if we would have a "Home" section like the [lsp-mode wiki](https://emacs-lsp.github.io/lsp-mode/). Then we could add a subsection similar to lsp-mode with Installation/Setup and the different possibilities. IMO that makes it a lot easier to get the information you need.

File: docs/nix-develop.md (hunk 8 above)
Comment (GTrunSec) [reply docs/nix-develop.md:29]: yes, it's necessary. I should explain the information to users on how to modify the build process by themselves. such as `configFlags` `patching the emacs-ng` `rust-nightly version`

**Comment by GTrunSec** [comment]
My purpose is to provide a more convenient and easy to modify `Nix-build` environment to help users custom the building options even in the develop mode.

**Comment by ericdallo** [comment]
> I think it would be cool if we would have a "Home" section like the lsp-mode wiki. Then we could add a subsection similar to lsp-mode with Installation/Setup and the different possibilities. IMO that makes it a lot easier to get the information you need.

I agree, this is possible changing the structure on mkdocs.yaml just like [lsp-mode mkdocs.yaml](https://github.com/emacs-lsp/lsp-mode/blob/master/mkdocs.yml#L7-L26)
c/c @GTrunSec

**Comment by GTrunSec** [comment]
> I agree, this is possible changing the structure on mkdocs.yaml just like [lsp-mode mkdocs.yaml](https://github.com/emacs-lsp/lsp-mode/blob/master/mkdocs.yml#L7-L26)

sorry about that, But I can move the current file under the path of `nix/`.

**Comment by ericdallo** [comment]
it needs to be under `docs` to mkdocs read that sucessfully

**Comment by GTrunSec** [comment]
> just like [lsp-mode mkdocs.yaml](https://github.com/emacs-lsp/lsp-mode/blob/master/mkdocs.yml#L7-L26)

Thanks, I got your point.

File: mkdocs.yml [hunk 9]
```diff
@@ -13,6 +13,7 @@ nav:
   - Getting started: getting-started.md
   - Using Deno: using-deno.md
   - Dynamic modules: ng-module.md
+  - Nix build: nix-develop.md
```
Comment (ericdallo) [inline mkdocs.yml:16]: Our horizontal navigation is already large: 
image

Following @brotzeit suggestion, we would need to change that to something like:

```yaml
nav:
    - Home:
      - Overview: README.md
      - Build / Develop: nix-develop.md
   ...
```

Comment (ericdallo) [reply mkdocs.yml:16]: Similar to [this](https://github.com/emacs-lsp/lsp-mode/blob/master/mkdocs.yml#L7-L26)

Comment (GTrunSec) [reply mkdocs.yml]: I see

**Comment by brotzeit** [comment]
Great job! I think we don't have to further improve the docs in this pull request, we can do this in a new one. There's already a volunteer in gitter who is waiting for this to be merged =)

File: mkdocs.yml [hunk 10]
```diff
@@ -6,7 +6,10 @@ repo_name: emacs-ng/emacs-ng
 repo_url: https://github.com/emacs-ng/emacs-ng
 
 nav:
-  - Home: README.md
+  - Home:
+    - Overview: README.md
+    - Setup:
+      - Nix Build / Develop: nix-develop.md
```
Comment (ericdallo) [inline mkdocs.yml:12]: Maybe in next PR move the getting started to another section of this `Setup`

**Review by ericdallo — APPROVED** [review]

**Comment by ericdallo** [comment]
Thank you for the time and help :)

**Pull request merged**